# Project 2
This will also be an exciting project, here we will **scrap the table, to make a table**. So, let's go. <br>
Here is the [link](https://pokemondb.net/pokedex/all)

# 

In [3]:
from selenium import webdriver
import pandas as pd
import numpy as np
import time
import warnings

warnings.filterwarnings("ignore")

In [11]:
site = "https://pokemondb.net/pokedex/all"
driver = webdriver.Edge("C:/WebDrivers/msedgedriver.exe")
driver.get(site)

# 

### Clicking on the Cookies button

In [12]:
time.sleep(2)
button = driver.find_element_by_xpath('//div[@id="gdpr-confirm"]/div/div/p[2]/button')
button.click()

### 

### Table!

In [13]:
table = driver.find_element_by_id('pokedex')

In [27]:
table_headers = table.find_element_by_xpath("thead/tr").text
table_headers = table_headers.split("\n")
table_headers

['#',
 'Name',
 'Type',
 'Total',
 'HP',
 'Attack',
 'Defense',
 'Sp. Atk',
 'Sp. Def',
 'Speed']

In [28]:
rows = table.find_elements_by_xpath("tbody/tr")

In [29]:
len(rows)

1047

There are `1047` rows to be scrapped... let's do it.

In [33]:
len(rows[0].find_elements_by_xpath("td"))

10

Which means, there are `10` columns on each row.

In [37]:
# Doing them in bulk
all_data = []
for row in rows:
    data = list(map(lambda x:x.text, row.find_elements_by_xpath("td")))
    all_data.append(data)

In [38]:
df = pd.DataFrame(all_data, columns=table_headers)
df

#                     Name            Type Total   HP Attack Defense  \
0     001                Bulbasaur   GRASS\nPOISON   318   45     49      49   
1     002                  Ivysaur   GRASS\nPOISON   405   60     62      63   
2     003                 Venusaur   GRASS\nPOISON   525   80     82      83   
3     003  Venusaur\nMega Venusaur   GRASS\nPOISON   625   80    100     123   
4     004               Charmander            FIRE   309   39     52      43   
...   ...                      ...             ...   ...  ...    ...     ...   
1042  896                Glastrier             ICE   580  100    145     130   
1043  897                Spectrier           GHOST   580  100     65      60   
1044  898                  Calyrex  PSYCHIC\nGRASS   500  100     80      80   
1045  898       Calyrex\nIce Rider    PSYCHIC\nICE   680  100    165     150   
1046  898    Calyrex\nShadow Rider  PSYCHIC\nGHOST   680  100     85      80   

     Sp. Atk Sp. Def Speed  
0         65      65    45  
1         80      80    60  
2        100     100    80  
3        122     120    80  
4         60      50    65  
...      ...     ...   ...  
1042      65     110    30  
1043     145      80   130  
1044      80      80    80  
1045      85     130    50  
1046     165     100   150  

[1047 rows x 10 columns]

In [39]:
driver.close()

### Cool! 
Now, we have the data... so let's clean it a bit and make it more analyzable.

In [45]:
df["Type"] = df["Type"].apply(lambda x:", ".join(x.split("\n")))

In [50]:
df["Total"] = df["Total"].astype(int)
df["HP"] = df["HP"].astype(int)
df["Attack"] = df["Attack"].astype(int)
df["Defense"] = df["Defense"].astype(int)
df["Sp. Atk"] = df["Sp. Atk"].astype(int)
df["Sp. Def"] = df["Sp. Def"].astype(int)
df["Speed"] = df["Speed"].astype(int)

In [57]:
def alt_name(name):
    names = name.split("\n")
    if len(names) > 1:
        return names[0] + " (" + names[1] + ")"
    return name

In [59]:
df["Name"] = df["Name"].apply(alt_name)

# Finally

In [62]:
df

#                      Name            Type  Total   HP  Attack  \
0     001                 Bulbasaur   GRASS, POISON    318   45      49   
1     002                   Ivysaur   GRASS, POISON    405   60      62   
2     003                  Venusaur   GRASS, POISON    525   80      82   
3     003  Venusaur (Mega Venusaur)   GRASS, POISON    625   80     100   
4     004                Charmander            FIRE    309   39      52   
...   ...                       ...             ...    ...  ...     ...   
1042  896                 Glastrier             ICE    580  100     145   
1043  897                 Spectrier           GHOST    580  100      65   
1044  898                   Calyrex  PSYCHIC, GRASS    500  100      80   
1045  898       Calyrex (Ice Rider)    PSYCHIC, ICE    680  100     165   
1046  898    Calyrex (Shadow Rider)  PSYCHIC, GHOST    680  100      85   

      Defense  Sp. Atk  Sp. Def  Speed  
0          49       65       65     45  
1          63       80       80     60  
2          83      100      100     80  
3         123      122      120     80  
4          43       60       50     65  
...       ...      ...      ...    ...  
1042      130       65      110     30  
1043       60      145       80    130  
1044       80       80       80     80  
1045      150       85      130     50  
1046       80      165      100    150  

[1047 rows x 10 columns]

In [63]:
df.to_csv("pokemon_scraped_data.csv", index=False)

# Done?
Yeah!! We are done with this basic course and also the project! 

**Things learnt:**
- A great flexibility by using Xpath
- Using selenium for **a real** kind of automation
- Knew that "we can click" without clicking!
- Grisp on selenium and syntax

Yeah! So I will see you Aayush in the next amazing course!